python/freetensor/core/transformer.py

```diff
@@ -253,35 +253,65 @@ def annotate_stmt(name: str, ty):
     return None
 
 
+class StagedPredicate(abc.ABC):
+
+    @abc.abstractmethod
+    def if_then_else_stmt(self, then_body: Callable[[], None],
+                          else_body: Optional[Callable[[], None]]):
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def if_then_else_expr(self, then_expr: Callable[[], Any],
+                          else_expr: Callable[[], Any]):
+        raise NotImplementedError()
+
+
+StagedPredicate.register(VarRef)
+
+
+def _varref_if_then_else_stmt(pred: VarRef, then_body: Callable[[], None],
+                              else_body: Optional[Callable[[], None]]):
+    with If(pred):
+        with LifetimeScope():
+            then_body()
+    if else_body:
+        with Else():
+            with LifetimeScope():
+                return else_body()
+
+
+def _varref_if_then_else_expr(pred: VarRef, then_expr: Callable[[], VarRef],
+                              else_expr: Callable[[], VarRef]):
+    return if_then_else(pred, then_expr(), else_expr())
+
+
+VarRef.if_then_else_stmt = _varref_if_then_else_stmt
+VarRef.if_then_else_expr = _varref_if_then_else_expr
+
+
 def if_then_else_stmt(predicate, then_body, else_body=None):
     '''If-then-else statement staging tool.
     When predicate is deterministic in staging, only one branch is generated.
     Otherwise, a If node in IR is generated.
     '''
-    if type(predicate) == bool:
+    if isinstance(predicate, StagedPredicate):
```

Comment:
What if `predicate` is neither a `StagedPredicate` nor a `bool`? I see a `VarRef` is now a `StagedPredicate`. What about a `ffi.Expr`? Is it a `StagedPredicate`?